proposals/reward-proposal-v2-reward-calculation.md

```diff
@@ -0,0 +1,109 @@
+---
+title: Epoch Reward Calculation
+---
+
+## Problem
+
+Calculating and distributing rewards at the first slot on the epoch boundary is
+not scalable. With increasing number of stake accounts and vote accounts on
+validator nodes, the block time at epoch boundary can last as long as 20 seconds
+because of paying the epoch rewards. Such a long block time degrades the network
+performance and becomes a potential vulnerability of the network.
+
+To solve this problem and speed up the reward time at epoch boundary, an earlier
+approach [Partitioned Inflationary Rewards
+Distribution](https://github.com/solana-labs/solana/blob/master/docs/src/proposals/partitioned-inflationary-rewards-distribution.md)
+is proposed. Let's call this approach `V1`. A prototype of the `V1` proposal is
+implemented. While the `V1` approach works functionally, it, however, has two
+main drawbacks.
+
+First, a restriction is enforced to prevent updates to all stake accounts and
+vote accounts during reward interval. In other words, the stake accounts and
+vote accounts are **locked** during reward interval. All stake account's actions
+such as delegate, redelegate, withdraw etc. are disallowed during reward
+interval and have to wait after reward interval.
+
+Second, there is no double-entry booking available for epoch rewards. Currently,
+epoch rewards are rewarded and recorded inside the first block of the epoch.
+Lack of double-entry booking is less a problem for the current approach, because
+all rewards are contained one block - the first block of the epoch, which makes
+block reward verification somewhat easier and trackable. However, with
+distributed rewards (such as the `V1` approach ), lack of double-entry booking
+becomes a more serious limitation. Since the rewards distribution can span
+multiple blocks, it becomes more difficult to verify the correctness of the
+reward payment without the ability for double-entry booking.
+
+Given these two drawbacks above, we are proposing an improved way to handle
+epoch reward - `V2`. The `V2` approach is based on `V1` but with the
+improvements to address these two drawbacks.
+
+To make the discussion clearer, the `V2` approach is split into two proposals:
+`rewards calculation` and `rewards distribution`. In this proposal, we will
+focus on "reward calculation". A follow-up proposal will talk about "reward
+distribution".
+
+## Proposed Solutions
+
+Unlike the current epoch rewards computation approach, in which computing the
+rewards happens at the first slot of epoch boundary in the bank runtime
+directly, the new reward computation is moved to a background service in the
+proposed approach.
+
+This is very similar to the `V1` proposal, a separate service,
+"EpochRewardCalculationService" is created at the start time of the validator.
+This service is listening on a channel for incoming rewards calculation
+requests, and handle the requests to compute the rewards.
```

Comment:
There are two options to handle cluster restart during reward interval. 

- Option 1: prevent taking snapshot and restarting from snapshot during
    reward/distributions.

      This will work in most cases. One challenge is how to handle cluster
      restart. For example, when a cluster fails at E+2 and E+1 is the slot that
      the cluster agreed upon, where E is the epoch start slot. A snapshot will be
      created for E+1. And all nodes on the cluster will restart from snapshot
      E+1. This is how cluster restart works currently. However, this won't work
      with reward interval. Instead, we need to restart from a snapshot at E-1 and
      let all nodes wait for supermajority on the hash of E+1, 2 slots ahead.
      Would the new restart behavior be acceptable?


    - Option 2: support taking snapshot and restart during reward
    calculation/distribution

      This is a more general approach. It requires adding extra information to
      the snapshots and modify bank replay during restart.

      - Snapshot generation:

         1. When taking snapshot reward_calculation, a snap of stake_cache at E
            will be serialized and stored in the snapshot.

         1. When taking snapshot reward_distribution, reward_calculation_result
            will be serialized and stored in the snapshot.

      - Restarting:

         1. When restarting from a snapshot during reward_calculation, a reward
            calculation request will be recreated from the snap of stake_cache
            at E, deserialized from the snapshot and sent to
            reward_calculation_service to restart the reward calculation.

         1. When restarting from a snapshot during reward_distribution, the
            reward calculation result will be loaded and deserialized from the
            snapshot. The merkle tree reward proof from the reward result will
            be recalculated. The recalculated reward calculation result and
            reward proof are used to distribute rewards.

Reply:
> * What is the relation between this service and different forks?
The calculation is per request signature not per fork. As long as all forks share the same request signature, there will be just one compuation.

> * How many threads does this service spawn per fork?
One.

> * How much overhead does this service add when there are excessive forking?
If all forks share the same root to E-1 slot, then there is only one compuation because the calculation request signatures are the same.